File: T1/2/t1_2.py
import bisect # This is for binary search

values = [int(x) for x in input().split()]
n = values[0]
q = values[1]
numbers = []
numbers = [int(x) for x in input().split()]

number_to_its_index = {} 
for i in range(n) : # I have dict from number to its indices+1
    number_to_its_index.setdefault(numbers[i], []).append(i + 1)

# get queries
for i in range(q):
    values = [int(x) for x in input().split()]
    r = values[0]
    l = values[1]
    if(r==l):
        print(numbers[r-1])
        continue
    
    half_length = (l-r+1)//2
    
    # checking the first number
    number = numbers[r-1]
    left_index = bisect.bisect_left(number_to_its_index[number], r)
    right_index = bisect.bisect_right(number_to_its_index[number], l)
    # Each bisect_left or bisect_right operation takes O(log n) time since it's a binary search.
    if(right_index - left_index > half_length):
        print(number)
        continue
    
    # checking the middle number
    number = numbers[r+((l-r-1)//2)]
    left_index = bisect.bisect_left(number_to_its_index[number], r)
    right_index = bisect.bisect_right(number_to_its_index[number], l)
    if(right_index - left_index > half_length):
        print(number)
        continue
    
    # If first number or middle number wasn't answer, so we don't have answer.
    else : print(0)
    # At first I want to implement random algorithm, after some checkings I came to this conclusion.

File: T1/2/t1_2.cpp
#include <algorithm> // This is for binary search (like bisect in python)
#include <iostream>
#include <random>
#include <unordered_map>
#include <vector>

using namespace std;

int main() {
    int n, q;
    cin >> n >> q;
    int numbers[n];
    unordered_map<int, vector<int>> number_to_its_indecis;
    for (int i = 0; i < n; i++) {
        int num;
        cin >> num;
        numbers[i] = num;
        number_to_its_indecis[num].push_back(i + 1);
    }

    // random index generatation
    random_device rd;
    mt19937 gen(rd());
    // Mersenne Twister (mt19937) is an excellent choice for random number generation
    // due to its high-quality randomness and fast performance.

    for (int i = 0; i < q; i++) {
        bool find = false;
        int l, r;
        cin >> l >> r;
        int half_len = (r - l + 1) / 2;
        uniform_int_distribution<> dis(l - 1, r - 1);
        // Uniform distribution(uniform_int_distribution<>) ensures that
        // each number between l and r has an equal probability of being selected.

        for (int j = 0; j < 17; j++) {
            int index = dis(gen);
            int number = numbers[index];
            vector<int> &indices = number_to_its_indecis[number];

            // Equivalent to bisect_left: Find the first position where index >= l
            auto left_it = lower_bound(indices.begin(), indices.end(), l);
            int left_index = left_it - indices.begin(); // Convert iterator to index

            // Equivalent to bisect_right: Find the first position where index < r
            auto right_it = upper_bound(indices.begin(), indices.end(), r);
            int right_index = right_it - indices.begin(); // Convert iterator to index

            if (half_len < right_index - left_index) {
                cout << number << "\n";
                find = true;
                break;
            }
        }
        if (!find) cout << 0 << "\n";
    }
}
